docs/hybridimages.md

```diff
@@ -0,0 +1,241 @@
+- A delegated dataset with 30 GiB of available space.
+- `/smartdc` needs to be a read-only lofs mount from the global zone
+- `/dev/kvm` device needs to be delegated for QEMU hardware acceleration
+- `git`, `gpg`, and `mkisofs` need to be installed and in `$PATH`.  If `pigz` is
+  available, the image creation will be quicker.
+
+Some images may have additional requirements.
+
+The following script may be used to customize an instance that was created with
+`triton instance create`.
+
+```bash
+#! /bin/bash
+
+uuid=$1
+if [[ -z $uuid ]]; then
+	echo "Usage: $0 <uuid>" 1>&2
+	exit 1
+}
+set -euo pipefail
+
+topds=zones/$uuid/data
+zfs create -o zoned=on -o mountpoint=/data $topds
+
+zonecfg -z $uuid <<EOF
+add dataset
+set name=$topds
+end
+add fs
+set dir=/smartdc
+set special=/smartds
+set type=lofs
+set options=ro
+end
+add device
+set match=kvm
+end
+EOF
+
+zlogin $uuid /opt/local/bin/pkgin update
+zlogin $uuid /opt/local/bin/pkgin -y install git gpg cdrtools pigz
+vmadm stop $uuid
+vmadm start $uuid
+```
+
+
+## Image builds
+
+The following repositories follow the patterns described above.
+
+* [mi-centos-hvm](https://github.com/joyent/mi-centos-hvm) for CentOS 6 - 8
+* [mi-debian-hvm](https://github.com/joyent/mi-debian-hvm) for Debian 8 - 10
+
+Each repo is intended to handle a family of distributions.  The mi-centos-hvm
+repo probably implements the hard parts required for RHEL, Fedora, and other
+RHEL-derived distributions.  Likewise, it should be straight-forward to add
+support for Ubuntu to  mi-debian-hvm.
+
+Each of those repositories has this ([eng](https://github.com/joyent/eng)) and
+[sdc-vmtools](https://github.com/joyent/sdc-vmtools) as subrepos.  The eng repo
+his home to common tools used for building the images and the sdc-vmtools repo
+is used for things that are added to the image.
+
+Some images may require software that is specific to the OS version and is
+stored elsewhere.  For example, some older versions ship with a version of
+cloud-init that is too old to properly support the metadata protocol.  On these
+systems, a custom build of cloud-init is stored in Manta and installed via
+post-install scripting that is invoked by the guest operating system installer
+during image creation.
+
+Each repository is designed to produce images using Jenkins, with the build
+steps for each image specified in the `Jenkinsfile` in each repository.
+
+
+## `create-hybrid-image` script
+
+The [`create-hybrid-image`](../tools/create-hybrid-iamge) script is typically
+invoked by an image-specific `create-image` script that resides in a
+`mi-<mumble>-hvm` repository.  See the usage message for details on options that
+are supported.
+
+
+### Who can run `create-hybrid-image`
+
+This script is designed to be able to be run as any user that has the `Primary
+Administrator` profile.  Usually it will be invoked as root, who has sufficient
+permissions in all but the oddest of circumstnaces.  To make the script usable
+by alice:
```

Comment:
I'm sure you'll do this, but it'd be good to check the build as a non-root user just in case.

An example I hit today in joyent-retro.git was where, when running that build script as non-root, using 'pfexec' to execute the script, and underneath an NFS mountpoint, copied files/directories to the proto area were retaining ownership of 'timf' rather than 'root', which resulted in hilarity because one of those copied directories was `/root/.ssh`, and as the new dir wasn't owned by 'root', sshd in the resulting image wouldn't let me login with my ssh key, **and** the build script went out of its way to set root's password to `NP`. Fun times.

Reply:
will do

Reply:
I have created an image as a non-root user (uid 1000) that has the primary administrator role.  I created an instance from that image, then inside the instance:

```
root@d10:~# find / -mount -user 1000
root@d10:~#
```